Edit Information CSS, custom platform, profile picture

byroncjy opened the pull request:
Updates EditInformation's CSS and adds custom platform functionality
Adjusts backend routes and tests to allow for this

Resolves #82
resolves #78

SivanC commented on back-end/platforms.js:
```diff
@@ -26,25 +26,30 @@ router.put('/:id/platforms', (req, res) => {
-      'Twitter',
-      'Github'
-    ]
+    // // Authorized platform values
```
If not needed then delete

byroncjy replied:
deleted

SivanC commented on back-end/platforms.js:
```diff
@@ -26,25 +26,30 @@ router.put('/:id/platforms', (req, res) => {
+
+    // Check if all elements in the array have only 'platform' and 'info' fields
+    // This will eventually be done with validator.js
+    const isValid = platformInformation.every(
```
We have platform schema for this actually, check my branch

byroncjy replied:
deleted

SivanC commented on back-end/platforms.js:
```diff
@@ -26,25 +26,30 @@ router.put('/:id/platforms', (req, res) => {
+    const isValid = platformInformation.every(
+      item =>
+        Object.keys(item).length === 2 && // Check if only two keys exist
+        item.hasOwnProperty('platform') &&
```
the fields are `name` and `value`

byroncjy replied:
yeah, this was based on the frontend request data

SivanC commented on front-end/src/EditInformation.js:
```diff
@@ -83,6 +83,16 @@ const EditInformation = () => {
+        const updatedPlatformInformationMap = [
+            ...platformInformationMap,
+			// add isCustom flag
+            { platform: "", info: "", isCustom: true}, 
```
won't this break your validator?

byroncjy replied:
it would but that is what the later code is for:
![image](https://github.com/agiledev-students-fall2023/4-final-project-qr-code-generator-app/assets/87872349/d903019c-bcb3-4ee7-991f-5718a79678c2)
it removes the isCustom field from the data sent to the backend, so it still passes validator

SivanC commented on front-end/src/EditInformation.js:
```diff
@@ -159,9 +169,10 @@ const EditInformation = () => {
+      // Also only extract out platform and info, not isCustom
+      const filteredPlatformInformationMap = platformInformationMap
+        .filter((item) => item.platform !== '' && item.info !== '')
+		.map(({ platform, info }) => ({ platform, info }))
```
What does this do?

byroncjy replied:
see above comment

SivanC reviewed (changes requested):
Pretty good, noticed that when you try to save a custom platform it will just revert to being the dropdown. Also, reason for not just puttiing "Custom Platform" in the dropdown? Seems like it would fix this issue. I don't think we need people to be able to change custom platform to other types via dropdown.

No need to add extra validation in the back-end, that's what the schemas are for, and the values have different names so that needs to be updated in the rest of the code.

CSS changes that would be good if you have time, not a priority since still looks usable on phone:
- Make all the things to touch bigger because fingers are imprecise (higher prio than everything else)
- - more padding between everything
- Center header
- Make name bigger/more central and pfp upload button smaller/out of the way more
- "Select platform" dropdown slightly wider than others
- Justify X button to right

byroncjy commented:
> Pretty good, noticed that when you try to save a custom platform it will just revert to being the dropdown. Also, reason for not just putting "Custom Platform" in the dropdown? Seems like it would fix this issue. I don't think we need people to be able to change custom platform to other types via dropdown.
> 
Changed accordingly
> Picture upload doesn't work, not sure if it is supposed to or not
> 
Fixed to work/integrate with backend
> No need to add extra validation in the back-end, that's what the schemas are for, and the values have different names so that needs to be updated in the rest of the code.
> 
Validation deleted
Field names changed
> CSS changes that would be good if you have time, not a priority since still looks usable on phone:
> 
> * Make all the things to touch bigger because fingers are imprecise (higher prio than everything else css-wise)
> * more padding between everything
> * Center header
> * Make name bigger/more central and pfp upload button smaller/out of the way more
> * "Select platform" dropdown slightly wider than others
> * Justify X button to right
>
Done some of these, will do more if have time tomorrow

byroncjy commented:
Note that I also added fixes for profile picture functionality plus integration with backend

XinranTaoAngela commented:
looks good!
